### PR TITLE
Use OEC approved language for SMS

### LINKED ISSUE #33: US # 7 Send Text When a New Document is Indexed
As the system, it wants to be notified when a new document is indexed in DocuClass so it can automatically send a text to the Participant confirming receipt of a new document

Acceptance Criteria
The system automatically checks periodically when the new document is indexed in DocuClass
AND THEN confirms Participant participation in the pilot
AND THEN automatically triggers a "Document Received" text to confirmed pilot participants
AND THEN the Participant receives the "Document Received" text

### PR DESCRIPTION
**Pull request by cailyncodes**
Should resolve #33. 

Uses OEC-approved language for document received event.

### CODE REVIEW

File: app/parent_notifications/notification_generator.rb [hunk 1]
```diff
@@ -16,8 +16,14 @@ def fetch_all_new
...
-      message_text: "We have received document #{event.type}"
+      message_text: build_document_received_message(
+        type: event.type, source: event.source, date: event.date
+      )
     )
     NotificationEvent.new(event.caseid, notification.id)
   end
+
+  def build_document_received_message(type:, source:, date:)
+    "This is Care 4 Kids! We have received your #{type} by #{source} on #{date.strftime('%m/%d/%Y')}. You will be notified if there is any missing information."
```
Comment (mmazanec22) [inline app/parent_notifications/notification_generator.rb]: Not sure if this is one to think about in this PR but it sounded from the email @geramirez forwarded us that type would be an abbreviation.  I would think that we want to make sure we're sending real words that are intelligible to people.  This might be something to revisit after we have db access?

Comment (cailyncodes) [reply app/parent_notifications/notification_generator.rb]: Oh good catch!! I completely forgot about that. Yeah, I think we should pause this PR until db access.

File: Dockerfile [hunk 2]
```diff
@@ -5,7 +5,7 @@ RUN apt-get update -qq && apt-get install -y build-essential \
     libqtwebkit4 \
     libqt4-dev xvfb \ 
     mariadb-client \
-    crontab \
+    cron \
```
Comment (geramirez) [inline Dockerfile]: why the change?

File: spec/parent_notifications/notification_generator_spec.rb [hunk 3]
```diff
@@ -18,29 +20,33 @@
...
+          time: time
         )
         notification_generator = NotificationGenerator.new(
           document_assigned_events: document_assigned_events
         )
 
         notification_events = fetch_all_new notification_generator
         stored_notification = Notification.last
-        expect(stored_notification.message_text).to eql 'We have received document some doc'
+        expect(stored_notification.message_text).to eql 'This is Care 4 Kids! We have received your some doc by Fax on 01/01/2010. You will be notified if there is any missing information.'
```
Comment (geramirez) [inline spec/parent_notifications/notification_generator_spec.rb]: This has nothing to do with the code, but how will they be notified? Via mail?

File: Dockerfile (hunk 2 above)
Comment (cailyncodes) [reply Dockerfile]: Neither Melanie nor I were able to install `crontab` -- apt didn't seem to think it was a package

File: spec/parent_notifications/notification_generator_spec.rb (hunk 3 above)
Comment (cailyncodes) [reply spec/parent_notifications/notification_generator_spec.rb]: Yeah I think it is through post mail

File: Dockerfile (hunk 2 above)
Comment (geramirez) [reply Dockerfile]: weird. no idea how that worked for me.

File: app/parent_notifications/document_assigned_events.rb [hunk 4]
```diff
@@ -50,14 +50,58 @@ def self.update_cursor(events:)
...
+    end
+    doc_type_parts[1].downcase
+  end
+
+  def self.parse_source_from_export(row)
+    source = row['Source']
+    if (!['Fax', 'Emails', 'Scans'].include?(source))
+      raise InvalidDocuclassDataTypeError
+    end
+    source.downcase
```
Comment (cailyncodes) [inline app/parent_notifications/document_assigned_events.rb]: This isn't it just quite. We need to match fax to fax, emails to uploads, and scans to mail.

File: spec/parent_notifications/document_assigned_events_spec.rb [hunk 5]
```diff
@@ -391,4 +391,81 @@
...
+        expect {
+          DocumentAssignedEvents.build_event(
+            {
+              'ClientID' => 1,
+              'DocType' => 'RP - Redetermination',
+              'Source' => 'Fax',
+              'ArchivedAt' => '2010-10-10 02:34:56'
+            }
+          )
+        }.to raise_error(InvalidDocuclassDataTypeError)
```
Comment (geramirez) [inline spec/parent_notifications/document_assigned_events_spec.rb]: I haven't run this myself so I'm not sure about this. Will raising an error break the program execution? I'm worried that if one of these errors is triggered, the cursor won't update and we'll get stuck in an infinite loop. Everything might be fine, I might just need an explanation.

File: app/parent_notifications/document_assigned_events.rb [hunk 6]
```diff
@@ -50,14 +50,58 @@ def self.update_cursor(events:)
...
+      parse_doc_type_from_export(row),
+      parse_source_from_export(row),
       parse_datetime_from_export(row)
     )
   end
 
+  def self.parse_client_id_from_export(row)
+    client_id = row['ClientID']
+    if (!client_id.kind_of? String)
+      raise InvalidDocuclassDataTypeError
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply app/parent_notifications/document_assigned_events.rb]: Hmm I don't think so. Thanks for catching this. I think I was thinking that the job processing would catch it and add it to our failed jobs table, but this is all stuff that happens before that. I'll add an e2e test to catch this and then fix the test.

File: app/parent_notifications/notification_generator.rb [hunk 7]
```diff
@@ -14,12 +14,25 @@ def fetch_all_new
...
     )
     NotificationEvent.new(event.caseid, notification.id)
   end
 
   def parent_is_active(parent)
     !parent.nil? and parent.active
   end
+  
+  def build_document_received_message(type:, source:, date:)
+    # Formats to single line
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply app/parent_notifications/notification_generator.rb]: 👍

File: app/parent_notifications/notification_generator.rb [hunk 8]
```diff
@@ -14,12 +14,25 @@ def fetch_all_new
...
+    <<-string.gsub(/\s+/, " ").strip
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply app/parent_notifications/notification_generator.rb]: Yeah it was super cool to come across

File: spec/parent_notifications/document_assigned_events_spec.rb [hunk 9]
```diff
@@ -391,4 +391,81 @@
       end
     end
   end
+
+  describe 'build_event' do
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply spec/parent_notifications/document_assigned_events_spec.rb]: Agreed! I didn't love doing this but wanted to make sure these were individually tested.

File: spec/parent_notifications/document_assigned_events_spec.rb [hunk 10]
```diff
@@ -391,4 +391,81 @@
...
+    context 'when parameters are valid types' do
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply spec/parent_notifications/document_assigned_events_spec.rb]: 👍

File: spec/parent_notifications/document_assigned_events_spec.rb (hunk 5 above)
Comment (cailyncodes) [reply spec/parent_notifications/document_assigned_events_spec.rb]: Not sure about the event cursor, will take a look, but yes see the above comment that the error needs to be addressed.

File: app/parent_notifications/document_assigned_events.rb [hunk 11]
```diff
@@ -50,14 +50,58 @@ def self.update_cursor(events:)
...
+    if (!['Fax', 'Emails', 'Scans'].include?(source))
+      raise InvalidDocuclassDataTypeError
+    end
+    source.downcase
+  end
+
   def self.parse_datetime_from_export(row)
-    Time.parse("#{row['ArchivedAt']}")
+    archived_at = row['ArchivedAt']
+    Time.parse("#{archived_at}")
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply app/parent_notifications/document_assigned_events.rb]: I realized that some of these we won't be able to test. ArchivedAt and Client ID in particular. The database column is formatted as a smalldatetime we can't write non-dates to it to test the response. And for Client ID anything will be converted into a string before writing it to the database. I removed those tests and simplified these methods.

File: spec/parent_notifications/document_assigned_events_spec.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (cailyncodes) [reply spec/parent_notifications/document_assigned_events_spec.rb]: Okay, so trying to test the error issue you raised was much more difficult than expected. In solving that, these tests disappeared so its a moot point now.

File: spec/parent_notifications/document_assigned_events_spec.rb (hunk 5 above)
Comment (cailyncodes) [reply spec/parent_notifications/document_assigned_events_spec.rb]: Okay so the solution I came up to potentially abuts this issue. Instead of raising an exception, it returns null and then we filter out the nulls. The update cursor method only updates if at least one event was returned. So if all of the data is invalid the cursor will not be updated but then I think we have a larger issue on our hands haha. What does raise an interesting question is how do we "save" the notifications that would be sent absent an invalid data type? Currently, we will be logging the error but don't track it anywhere. Thoughts?

**Comment by cailyncodes** [comment]
Hmm these tests are passing locally. Any ideas?

**Comment by cailyncodes** [comment]
Omg debugging this was an absolute nightmare (sorry to the circle ci jobs logs). In short, there were multiple event_cursor objects being inserted into the table and for some reason locally it was using the most recent one and on CircleCI it was using the least recent one. I added a `EventCursor.delete_all` into the integration test and it fixed everything.

File: spec/parent_notifications/document_assigned_events_spec.rb [hunk 12]
```diff
@@ -391,4 +391,107 @@
       end
     end
   end
+
+  context 'when an event has all database fields that are valid types' do
```
Comment (geramirez) [inline spec/parent_notifications/document_assigned_events_spec.rb]: Since we have a couple of days, I feel we can move this method out of the `document_assigned_events.rb` file and into it's own class. Maybe in a new PR.

**Comment by geramirez** [comment]
LGTM, 
This PR also raises a couple of chores we need to take care of.
1. a unique constraint on our cursor table (I'm so glad you ran into this issue)
2. moving the build_event logic into it's own class so it's easier to test.

Number 1 is a lot more important. Let's figure out how to split this up later.
